File: src/test/java/com/mpe85/grampa/runner/DefaultParseRunnerTest.java
package com.mpe85.grampa.runner;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.Test;

import com.mpe85.grampa.event.ParseEventListener;
import com.mpe85.grampa.event.PreMatchEvent;
import com.mpe85.grampa.parser.AbstractParser;
import com.mpe85.grampa.rule.Rule;
import com.mpe85.grampa.rule.impl.EmptyRule;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

@SuppressFBWarnings(value = "SIC_INNER_SHOULD_BE_STATIC_ANON", justification = "Performance is not of great importance in unit tests.")
public class DefaultParseRunnerTest {
	
	private static final class IntegerTestListener extends ParseEventListener<Integer> {
		@Override
		public void beforeMatch(final PreMatchEvent<Integer> event) {
			throw new RuntimeException();
		}
	}
	
	@Test
	public void getRootRule() {
		final class Parser extends AbstractParser<Integer> {
			@Override
			public Rule<Integer> root() {
				return empty();
			}
		}
		final DefaultParseRunner<Integer> runner = new DefaultParseRunner<>(new Parser());
		assertTrue(runner.getRootRule() instanceof EmptyRule);
	}
	
	@Test
	public void registerListener() {
		final class Parser extends AbstractParser<Integer> {
			@Override
			public Rule<Integer> root() {
				return empty();
			}
		}
		final DefaultParseRunner<Integer> runner = new DefaultParseRunner<>(new Parser(), (ex, ctx) -> {
			assertTrue(ex instanceof RuntimeException);
		});
		final ParseEventListener<Integer> listener = new IntegerTestListener();
		runner.registerListener(listener);
		assertDoesNotThrow(() -> runner.run("a"));
	}
	
	@Test
	public void unregisterListener() {
		final class Parser extends AbstractParser<Integer> {
			@Override
			public Rule<Integer> root() {
				return empty();
			}
		}
		final DefaultParseRunner<Integer> runner = new DefaultParseRunner<>(new Parser(), (ex, ctx) -> {
			fail("Listener should not have been called.", ex);
		});
		final ParseEventListener<Integer> listener = new IntegerTestListener();
		runner.registerListener(listener);
		runner.unregisterListener(listener);
		runner.run("a");
	}
	
}

File: src/test/kotlin/com/mpe85/grampa/runner/DefaultParseRunnerTest.kt
package com.mpe85.grampa.runner

import com.mpe85.grampa.event.ParseEventListener
import com.mpe85.grampa.event.PreMatchEvent
import com.mpe85.grampa.grammar.AbstractGrammar
import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.rule.impl.EmptyRule
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

@SuppressFBWarnings(
  value = ["SIC_INNER_SHOULD_BE_STATIC_ANON"],
  justification = "Performance is not of great importance in unit tests."
)
class DefaultParseRunnerTest {

  private class IntegerTestListener : ParseEventListener<Int>() {
    override fun beforeMatch(event: PreMatchEvent<Int>) {
      throw RuntimeException()
    }
  }

  @Test
  fun getRootRule() {
    class Grammar : AbstractGrammar<Int>() {
      override fun root(): Rule<Int> {
        return empty()
      }
    }

    val runner = DefaultParseRunner(Grammar())
    assertTrue(runner.rootRule is EmptyRule<*>)
  }

  @Test
  fun registerListener() {
    class Grammar : AbstractGrammar<Int>() {
      override fun root(): Rule<Int> {
        return empty()
      }
    }

    val runner = DefaultParseRunner(Grammar())
    val listener = IntegerTestListener()
    runner.registerListener(listener)
    assertDoesNotThrow<ParseResult<Int>> { runner.run("a") }
  }

  @Test
  fun unregisterListener() {
    class Grammar : AbstractGrammar<Int>() {
      override fun root(): Rule<Int> {
        return empty()
      }
    }

    val runner = DefaultParseRunner(Grammar())
    val listener = IntegerTestListener()
    runner.registerListener(listener)
    runner.unregisterListener(listener)
    runner.run("a")
  }

}
